common/src/chain/transaction/output/htlc.rs

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) 2024 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// TODO: consider removing this in the future when fixed-hash fixes this problem
+#![allow(clippy::non_canonical_clone_impl)]
+
+use std::time::Duration;
+
+use serialization::{Decode, Encode};
+use zeroize::Zeroize;
+
+use super::Destination;
+
+#[derive(Debug, Clone, PartialEq, Eq, Encode, Decode, serde::Serialize, serde::Deserialize)]
+pub struct HashedTimelockContract {
+    // can be spend either by a specific address that knows the secret
+    pub secret_hash: HtlcSecretHash,
+    pub spend_key: Destination,
+
+    // or by a multisig after timelock expires making it possible to refund
+    pub refund_timelock: Duration,
+    pub refund_key: Destination,
+}
+
+#[derive(Zeroize, Clone, Encode, Decode, PartialEq, Eq, Debug)]
+pub struct HtlcSecret {
+    pub secret: [u8; 20],
+}
+
+fixed_hash::construct_fixed_hash! {
+    #[derive(Encode, Decode)]
+    pub struct HtlcSecretHash(20);
+}
+
+impl rpc_description::HasValueHint for HtlcSecretHash {
+    const HINT_SER: rpc_description::ValueHint = rpc_description::ValueHint::HEX_STRING;
+}
+
+impl serde::Serialize for HtlcSecretHash {
+    fn serialize<S: serde::Serializer>(&self, s: S) -> Result<S::Ok, S::Error> {
+        s.serialize_str(&format!("{self:x}"))
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for HtlcSecretHash {
+    fn deserialize<D: serde::Deserializer<'de>>(d: D) -> Result<Self, D::Error> {
+        struct HashVisitor;
+        impl<'de> serde::de::Visitor<'de> for HashVisitor {
+            type Value = HtlcSecretHash;
+            fn expecting(&self, fmt: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                fmt.write_str("a hex-encoded hash")
+            }
+            fn visit_str<E: serde::de::Error>(self, s: &str) -> Result<Self::Value, E> {
+                s.parse().map_err(serde::de::Error::custom)
+            }
+        }
+        d.deserialize_str(HashVisitor)
```

Comment:
Why not simply expect hex and deserialize from hex?

Reply:
I mean, isn't this what you're doing in the serialize function?

Reply:
It actually does it via FromStr implementation for fixed hash. I took it from H256

Reply:
Please add tests below that serialize and deserialize, both in the cases of success and error.

Reply:
Also, please add fixed-cases, to ensure that upgrades won't mess this up.